### PR TITLE
ci, packaging, meta: modernize Python packaging for Sopel

### PR DESCRIPTION
**Pull request by Exirel**
### Description

After hours of reading various documentations (mostly: the [packaging user guide](https://packaging.python.org/en/latest/) and [setuptools guide for pyproject.toml](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html)), I think I end up with a pretty good understanding of the topic matter to modernize Sopel's build system. I also read various PEP, such as [PEP 517](https://peps.python.org/pep-0517/), [PEP 518](https://peps.python.org/pep-0518/), [PEP 643](https://peps.python.org/pep-0643/), [PEP 660](https://peps.python.org/pep-0660/), with various results.

#### How to review this PR

You have various options: I advise to, at least, try out a `pip install -e .` locally to see if it works for you.
The best thing would be to read all the links and do some research to have all the knowledge needed, but since it might take a lot of your time, I tried to write down all the important bit in this PR's description.

Take your time to read it fully, and ask me questions on IRC.

#### Now and then

At the moment, Sopel's build system is composed of the following steps:

* local install is done through either `pip install -e .` or `python setup.py develop`, then requires `pip install -U -r dev-requirements.txt`
* build the source and built distribution with `python setup.py sdist bdist`, generating a `.tar.gz` and a `.whl`
* upload with `twine`
* user install is done through `pip install sopel`

The goal of this PR is to switch to the more modern and standard approach to build package in Python:

* local install is done through `pip install -e .`, then requires `pip install -U -r dev-requirements.txt`
* build the source and built distribution with `python -m build`, generating a `.tar.gz` and a `.whl`
* upload with `twine`
* user install is done through `pip install sopel`

The workflow doesn't change (dev, build, upload, install), and the tools used for development and build are more standardized. Curious reader can have a look at the [official tutorial for Python packaging](https://packaging.python.org/en/latest/tutorials/packaging-projects/), which contains the same workflow and tools.

#### Switching to pyproject.toml

Technically, we don't **need** to switch from `setup.py` to `pyproject.toml` in order to use modern tools, such as `build`. However, the modern packaging is now standardized around that file, and there is no way around it: `setup.py` is an outdated tools that served us well for almost two decades ([distutils was created around 2000 and setuptools 2004](https://www.pypa.io/en/latest/history/)) and need to be retired.

The `pyproject.toml` contains two important sections:

* `[build-system]`, following PEP 517 and 518, to declare which build backend must be used to generate source and built distribution
* `[project]`, following [PEP 621](https://peps.python.org/pep-0621/) and [official documentation](https://packaging.python.org/en/latest/specifications/declaring-project-metadata/), to declare most metadata from setup.cfg/py into a normalized format

I decided to **keep using setuptools** as our backend of choice: this is the default backend used by `build`, ensuring that we don't switch to a complete new set of tools and configuration files. I ensured that everything was also available in `pyproject.toml`, with minor adjustments:

* finding the appropriate package with setuptools through `[tool.setuptools.packages.find]` (backend specific)
* remaining metadata still present through `[tool.setuptools]`
* `requirements.txt` is replaced by the now standard `project.dependencies` option
* entrypoints (both console and custom) are properly defined
* `[project.urls]` now contains more than the homepage

#### Removing setup.py, keeping setup.cfg

As a result of using `pyproject.toml` with the `setuptools` backend, I was able to completely remove the `setup.py` file. However, since we are still using `setuptools`, I was forced to keep the `setup.cfg` file. Setuptools's documentation points out that one may be forced to keep an empty `setup.py` file, or that `setuptools` will generates it if you have a `setup.cfg` file. Confusing? Yes. I know, `setuptools` is improving by the minute, so it's a small price to pay to be on track with the future of Python packaging.

I don't think it is relevant to perform various Python check ourselves: let the installer of the user do that for us. The Python requirement is now standard, so one should not be able to install Sopel with an outdated version of Python.

The `setup.cfg` file that's left is now empty of project's metadata, and contains only configurations for our development tools, such as `flake8` or `mypy`.

As a result, a developer will need to issue the following command to work on Sopel locally: `pip install -e . && pip install --upgrade -r dev-requirements.txt`.

Installing the project from source still works by following Python standard: `pip install .`.

#### About dependencies

There is a point that I'm not completely happy with: the requirements. I feel like tooling around them is not really at the top. I guess that's why PyPA suggest to use other tools than `pip` to develop projects, such as Hatch or Poetry. My goal with this modernization of our build system is to allow developers to use these tools if they want to, as long as we keep building it the same way for everyone.

In any case, I made two choices:

* remove the `requirements.txt` file, and put the requirements into the `pyproject.toml` file
* keep the `dev-requirements.txt` file, and not have a "dev requirements"  into the `pyproject.toml` file

I need to be able to do `pip install --upgrade -r dev-requirements.txt` to get the up to date version of flake8, pytest, and so on, because doing `pip install --upgrade -e .[dev]` didn't provide that: it would keep the currently installed version as long as they conformed to the spec. Other tools than pip can do that, but it didn't seem like a good idea to change that process for now.

#### Updating CI configuration

This is probably the most controversial change of this PR: I decided to update the CI to reflect modern standards. First thing first, I switch to `python -m build`, as recommended.

Then, I upgraded the github actions to the latest version of each.

And last but not least, I decided to use [PyPA's action to publish on PyPI](https://github.com/pypa/gh-action-pypi-publish).

As far as I know, everything works like a charm.

#### Removing old files

Both `sopel.py` and `pytest_run.py` are outdated tools. I don't understand why we kept them for so long, and what purpose they serve exactly. In any case, I removed them.

I updated the commit-hooks with something that should work. And if people wants to run Sopel from source, then they need to understand how they are supposed to do that: since it is now all standard and well documented, there is no reason to maintain anything more than the standard.

#### Improving CONTRIBUTING.md

I rewrite a part of CONTRIBUTING.md to explain with more details the typical development workflow for anyone wanted to provide a PR to the project: how to install, how to manage git, and how to run tests.

### Optional: authors & maintainers metadata

I don't know what to do with these fields. Maybe for another PR?

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:55]: I _think_ American spellings are pretty consistent throughout the docs. Not necessarily a preference, but I don't think anyone wants to go through all the documentation and change _everything_ to the UK spellings.

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:61]:
```suggestion
* Create a virtualenv.
  * You can use `virtualenv` directly, `virtualenvwrapper`, or any tool that
    allows you to create and manage your virtualenv.
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:67]:
```suggestion
* Install Sopel from source as an editable install.
  * If you don't use a project manager (like Poetry or Hatch), you can do
    that with `pip install -e .` from within the clone's directory, where
    there is a `pyproject.toml` file.
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:68]:
```suggestion
* Install development tools used to run tests and linters; there is a specific
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:73]:
```suggestion
At this point, you are all set, at least for the Python environment; all
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:74]:
```suggestion
that's left is for you to configure your code editor the way you like, and read and
```

I'm intentionally not making this suggestion bigger because I'm on the fence about "and read and write". Need another brainwave.

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:78]:
```suggestion
Using branches
--------------
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:81]:
```suggestion
changes and commit them. Moreover, we advise you to work from your own branch.
```

Additionally, could be useful to specify that each change/feature should be in its own branch.

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:111]: What is meant by "Said working", so I can suggest a more grammatically correct alternative to this line?

File: MANIFEST.in
Comment (dgw) [inline MANIFEST.in:1]: Was it bad that we were including `dev-requirements.txt` in the package before? 😁

File: conftest.py
Comment (dgw) [inline conftest.py:6]: Can we just remove this, since it's empty, or will it break something?

File: contrib/githooks/main-hook.sh
Comment (dgw) [inline contrib/githooks/main-hook.sh:72]: Is this truly equivalent to what was there before? Does running `pytest` this way include `@plugin.example()`s from core plugins?

File: pyproject.toml
Comment (dgw) [inline pyproject.toml:18]: Should list more than just me. Many people were responsible for creating the code that eventually became Sopel, then for maintaining Sopel before I took over as maintainer, and now for helping me maintain it (including you).

File: pyproject.toml
Comment (dgw) [inline pyproject.toml:49]: This is good. Expanding the listed URLs was on one of my informal to-do lists, actually.

Can we put GitHub Sponsors and Open Collective on here as "funding"/"donation" type? (https://github.com/pypi/warehouse/blob/01f87cfdb6275a8c8bc11b838618df246fc10fec/warehouse/templates/packaging/detail.html#L30-L31)

**Review by dgw — COMMENTED** [review]
Not possible to call out in line notes: The commit message for c686d51aa173f9e4d5c59bc7b2274ce5afe85c4a doesn't make sense. Shouldn't that be something like `ci: add wheel to initialization`?

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:55]: Oh, it's "favorite", right? You'll have to point me out these things, because I'm not sure I'm even aware of all of them.

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:111]: Said working... copy? Yeah, I think it's working copy. The goal is to talk about the local repository.

File: conftest.py
Comment (Exirel) [reply conftest.py:6]: We could probably remove it, yes.

File: contrib/githooks/main-hook.sh
Comment (Exirel) [reply contrib/githooks/main-hook.sh:72]: Yes, hence the `sopel` and `test/`. That's how I run tests locally.

File: pyproject.toml
Comment (Exirel) [reply pyproject.toml:49]: Absolutely!

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:55]:
```suggestion
the code using standard tooling, and take advantage of your favorite code
```

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:68]: You can see me try. :upside_down_face:

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:111]:
```suggestion
Said branch can be updated with rebase:
```

File: MANIFEST.in
Comment (Exirel) [reply MANIFEST.in:1]: Unless we want to run tests from source (but I don't see the point), there is no need for that, no.

**Comment by Exirel** [comment]
All done. I've squashed the weird commit and pushed additional commits (without a rebase on master this time).

> Should list more than just me. Many people were responsible for creating the code that eventually became Sopel, then for maintaining Sopel before I took over as maintainer, and now for helping me maintain it (including you).

That's a tricky question actually. The field "authors" isn't well specified, and no one seems to agree on what it actually means. Is it a list of active authors? A list of historical authors? A list of all the contributors? And how does "maintainers" fit into that?

What I suggest for now is to change that in a future PR, as this can be changed before the release. Right now, I don't have a meaningful opinion about that field.

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:76]: Line _looks_ too long (but I didn't count columns, just compared to above).

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:91]: For a subfolder named "sopel", this can be accomplished in fewer steps by:


```suggestion
$ cd path/to/your/dev/folder
$ git clone git@github.com:<USERNAME>/sopel.git
$ cd sopel/
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:106]: Why not `git merge --ff-only upstream/master`? Unless I'm way off base about how rebase works, using it here may change SHAs on the downstream fork because `pick`ing a commit _copies_ it (and applies a new committer date). I've never seen someone suggest using `rebase` for this task before. 🤔

```suggestion
$ git merge --ff-only upstream/master
```

True, a straight merge risks adding a merge commit, but `--ff-only` aborts if the merge isn't a fast-forward.

Then again, if the user truly hasn't touched their own master branch, the nuclear option is fine too 😁:

```suggestion
$ git reset --hard upstream/master
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:123]: Ach.

```suggestion
In both cases, if you configured
```

File: pyproject.toml
Comment (dgw) [reply pyproject.toml:18]: Let's make this its own issue, but not necessarily for the 8.0.0 milestone if how to structure the metadata about past contributors isn't clear yet.

File: pyproject.toml
Comment (dgw) [inline pyproject.toml:52]:
```suggestion
"Funding on Open Collective" = "https://opencollective.com/sopel"
"Sponsor on GitHub" = "https://github.com/sponsors/sopel-irc"
```

This will use the correct type of icon on PyPI for both links. It's chosen by matching what the string starts with (I know, I know, not my architectural decision).

It also differentiates the two a bit. The primary purpose of OC is public expenditure records, while all of our current sponsors do so via GH.

**Review by dgw — CHANGES_REQUESTED** [review]
Phone review today, but still caught some stuff. 💪

You're right about leaving the author etc. metadata for the future. See comment added to existing thread about that in TOML file.

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:106]: Ha ah. I do `git rebase upstream/master` all the time, because if you don't commit on your master branch, then what rebase does it take all the commit from upstream/master, and apply all the commits that exists only in your version. Therefore, only *your* commits are changed.

Basically, it's like `merge --ff-only`.

Comment (dgw) [reply CONTRIBUTING.md:106]: That'll count as "way off base about how rebase works", then. 😅

**Comment by Exirel** [comment]
I'll let you open the issue for the list of authors/maintainers. Other than that, everything else have been applied.

I'm ready to squash.

File: CONTRIBUTING.md
Comment (SnoopJeDi) [inline CONTRIBUTING.md:61]: May want to also mention `venv` here for a stdlib option, and a link to the [PyPA](https://packaging.python.org/en/latest/guides/installing-using-pip-and-virtual-environments/#creating-a-virtual-environment) guide might be helpful for contributors who aren't already familiar.

File: pyproject.toml
Comment (SnoopJeDi) [inline pyproject.toml:2]: I believe you could declare dependency on `wheel` here as well and strike the separate instruction to install that package before building/installing Sopel

File: CONTRIBUTING.md
Comment (dgw) [reply CONTRIBUTING.md:61]: Hm, the PyPA guide says "virtualenv (for Python 2)" 🤔

Are `virtualenv` and `venv` close enough to feature parity that we can just recommend the stdlib option (`venv`) for newbies "or any other appropriate tool" if someone already has a preferred one?

File: pyproject.toml
Comment (dgw) [reply pyproject.toml:2]: Good point. I do not know why `wheel` hasn't already been added as an explicit dependency. Seems like I've already run into needing to install that manually when releasing some package version or another, maybe while Travis was broken.

File: pyproject.toml
Comment (dgw) [reply pyproject.toml:18]: Done in #2339. Just the new comments from @SnoopJeDi left now.

File: CONTRIBUTING.md
Comment (SnoopJeDi) [reply CONTRIBUTING.md:61]: The main difference is that `venv` provides an environment for whatever Python it's running in, and `virtualenv` has more ambition for managing multiple versions.

I would definitely be inclined towards recommending "`venv` or other tools" here, the users who want something other than the standard option probably don't need the docs to guide them to it anyway.

Comment (Exirel) [reply CONTRIBUTING.md:61]: I'd rather say:

> You can use `virtualenv` directly (or the `venv` built-in), `virtualenvwrapper` to manage multiple projects, or any tool that [...]

**Comment by Exirel** [comment]
All good now. I tried to find a proper way to talk about venv and virtualenvs, and then I noticed that I should put a line about project manager tools that would create venv for developers.

Anyway, it should be good for a last round of review, and a squash. Then, I can start building Python 3.10 in the CI with that!

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:64]: Comma splice bad.

```suggestion
  * Your project manager may create the virtualenv for you; be sure to check
```

File: CONTRIBUTING.md
Comment (dgw) [inline CONTRIBUTING.md:68]: Is explicitly installing `wheel` still necessary? It and `setuptools` are declared as `build-system` dependencies in `pyproject.toml`.

File: pyproject.toml
Comment (dgw) [reply pyproject.toml]: This was resolved, but the separate instruction still exists. (See other line note in CONTRIBUTING.)

**Review by dgw — COMMENTED** [review]
I think there's just one loose end to tie off, and a tiny grammar thing.

You can squash at the same time. GH will let me see the force-push diff, which will be enough unless you also need to rebase for some reason.

File: CONTRIBUTING.md
Comment (Exirel) [reply CONTRIBUTING.md:68]: `wheel` is never necessary, it is only convenient: if installed, it means that doing `pip install -r dev-requirements.txt` for example will be able to generate wheels for the downloaded packages that don't provide them, even tho `pip install -e .` wasn't done yet.

So yeah, I consider that good advice anyway! It's like upgrading `pip` and installing `pytz`: these should be almost automatic when starting a new virtualenv.

**Comment by Exirel** [comment]
Grammar: fixed.
Comment: replied.
Commits: squashed.

**Review by dgw — APPROVED** [review]
I don't see any files unexpectedly marked as unviewed, so let's call it done. Probably want to test this one as a last sanity check before merging, when I'm at a computer later… 😅

**Comment by dgw** [comment]
Right, I went through the `python3 -m build` process and uploaded the results to testpypi, to see how the new metadata works. Few weird things that are easy to fix. I will push a few commits that fix these, momentarily, but here's the written explanation of what I changed and why (for later, when these are squashed into the right places in the PR history).

<details>
<summary>Full text of the license file is displayed in the sidebar.</summary>

Before:
image

After:
image

</details>

This one can be fixed by just using the "text" option to specify "Eiffel Forum License v2.0", similar to how the 7.x releases have. Put a pin in [PEP 639](https://peps.python.org/pep-0639/#core-metadata) (deprecates/replaces `License` and the license-related classifiers) for later.

I tried directly using the `license = "EFLv2"` format, but it's not supported yet. So we'll have to take another modernization pass at the metadata once the deprecations and their replacements actually _take effect_ I guess. (Attempt to future-proof rejected!) This is also why I haven't removed the license classifiers.

<details>
<summary>Funding links are separated</summary>

Old:
image

New:
image

</details>

This is because Warehouse/PyPI alphabetizes the project links. I would suggest using "Donate" or "Sponsor" for both GH and OC so they appear together in the list. Using "Donate" lets "Source" appear last, which I think is nice.

----

I also changed "Release notes" to point to https://sopel.chat/changelog/ (can't be seen in a screenshot, lol).

----

Also note, I often saw a warning in the output of `build` operations:

```
/tmp/build-env-s4nd2h1_/lib/python3.9/site-packages/setuptools/config/pyprojecttoml.py:108: _BetaConfiguration:
Support for `[tool.setuptools]` in `pyproject.toml` is still *beta*.
  warnings.warn(msg, _BetaConfiguration)
```

It's probably fine? But worth mentioning here, so future-us can see that it _was_ noticed while testing this patch. Yes, the warning is still there on setuptools' `main` branch.

**Comment by dgw** [comment]
Merging this broke unstable docs—whoops!—but the fix was mostly on the Netlify configuration side, out of tree. f2712b1033e231b19383fbdf895bfef153e53e76 contains the only repo tweak needed to make it work again.

**Pull request merged**
